### PR TITLE
Make some Mappers Singletons

### PR DESCRIPTION
**Pull request by original-brownbear**
Just some obvious singletons we can use to save a little memory/cache
that I found during experiments.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java [hunk 1]
```diff
@@ -93,21 +93,28 @@ public FieldNamesFieldMapper build() {
...
     public static final class FieldNamesFieldType extends TermBasedFieldType {
 
+        private static final FieldNamesFieldType ENABLED = new FieldNamesFieldType(true);
+
+        private static final FieldNamesFieldType DISABLED = new FieldNamesFieldType(false);
+
         private final boolean enabled;
 
-        public FieldNamesFieldType(boolean enabled) {
+        public static FieldNamesFieldType of(boolean enabled) {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java:113]: Maybe make this `get`, to align with `RoutingFieldMapper`? And does it need to be public?

**Review by romseygeek — APPROVED** [review]
LGTM, thanks @original-brownbear!

**Comment by original-brownbear** [comment]
Thanks @romseygeek !

File: server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java]: ++ renamed, we need this exposed for some tests that's why it's `public`.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 2]
```diff
@@ -385,8 +385,10 @@ protected void doXContentBody(XContentBuilder builder, Params params) throws IOE
 
     public static class MultiFields implements Iterable<FieldMapper>, ToXContent {
 
+        private static final MultiFields EMPTY = new MultiFields(Collections.emptyMap());
+
         public static MultiFields empty() {
-            return new MultiFields(Collections.emptyMap());
+            return EMPTY;
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:391]: Nice catch!

**Pull request merged**